### ISSUE TITLE
SNMPd sometimes sends SIGPIPEs to Clearwater processes

### ISSUE THREAD

**Issue by graemerobertson** [Metaswitch/clearwater-net-snmp#11]
#### Symptoms
Sprout was just dying silently on start-up. This turned out to be because it was failing to establish a connection with SNMPd, and SNMPd was sending it a SIGPIPE which we never expect Sprout to have to handle.

#### Impact
Initially struggled to debug an actual problem.

#### Release and environment 
net-snmp Version: 5.7.2~dfsg-clearwater5

#### Steps to reproduce
Seen due to https://github.com/Metaswitch/clearwater-net-snmp/issues/10.

**Comment by graemerobertson** [comment]
I'm not 100% sure this issue makes sense, but my understanding from conversations I've had today is that we should be stopping SNMPd from sending SIGPIPEs - @mirw please can you review/comment?

**Comment by rkd-msw** [comment]
I don't think snmpd is sending the SIGPIPE here - I think the SIGPIPE happens when Sprout writes to a UNIX domain socket which is closed (http://stackoverflow.com/questions/8369506/why-does-sigpipe-exist).

**Comment by graemerobertson** [comment]
Ah yes, ok, I was wondering why the strace for SNMPd wasn't showing anything about sending a SIGPIPE; does that mean that this issue is a) impossible to fix and/or b) in the wrong repo?

**Comment by mirw** [comment]
Agreed.  You can suppress `SIGPIPE` by calling `setsockopt(SO_NOSIGPIPE)` on the socket.  I think we should understand why snmpd is behaving in this way, but in the short term maybe we should do this in sprout/cpp-common (it's a difficult decision because I don't think we want sprout to die if snmpd dies, but we also have no way of subsequently recovering).

**Comment by rkd-msw** [comment]
Or catch SIGPIPE and do our normal die-in-ten-minutes exception handling?

**Comment by mirw** [comment]
Not sure.  It feels a bit scary catching all SIGPIPEs, but maybe.
